lib/interfaces/include/BMSDriverGroup.tpp

```diff
@@ -0,0 +1,571 @@
+                {
+                    max_humidity = gpio_in;
+                    bms_data.max_board_temperature_segment_id = (chip + 2) / 2;
+                }
+            }
+            else if (gpio_Index < 4)
+            {
+                float thermistor_resistance = (2740 / (gpio_in / 50000.0)) - 2740;
+                bms_data.cell_temperatures[gpio_count] = 1 / ((1 / 298.15) + (1 / 3984.0) * log(thermistor_resistance / 10000.0)) - 273.15; // calculation for thermistor temperature in C
+                total_thermistor_temps += gpio_in;
+                if (gpio_in > max_thermistor_voltage)
+                {
+                    max_thermistor_voltage = gpio_in;
+                    bms_data.max_cell_temperature_cell_id = gpio_count;
+                }
+            }
+            gpio_count++;
+        }
+    }
+
+    bms_data.min_voltage = this->min_voltage;
+    bms_data.max_voltage = this->max_voltage;
+    bms_data.total_voltage = this->total_voltage;
+    bms_data.average_cell_temperature = this->total_thermistor_temps / gpio_count;
+    return bms_data;
+}
+
+template <size_t num_chips, size_t num_chip_selects>
+void BMSDriverGroup<num_chips, num_chip_selects>::_reset_voltage_data()
+{
+    total_voltage = 0;
+    max_voltage = 0;
+    min_voltage = 65535;
+}
+
+template <size_t num_chips, size_t num_chip_selects>
+void BMSDriverGroup<num_chips, num_chip_selects>::_reset_GPIO_data()
+{
+    max_humidity = 0;
+    max_thermistor_voltage = 0;
+    max_board_temp_voltage = 0;
+    min_board_temp_voltage = 65535;
+    total_thermistor_temps = 0;
+}
+
+/* -------------------- WRITING DATA FUNCTIONS -------------------- */
+
+template <size_t num_chips, size_t num_chip_selects>
+void BMSDriverGroup<num_chips, num_chip_selects>::_write_configuration(uint8_t dcto_mode, const std::array<uint16_t, num_chips> &cell_balance_statuses)
+{
+    std::array<uint8_t, 6> buffer_format; // This buffer processing can be seen in more detail on page 62 of the data sheet
+    buffer_format[0] = (gpios_enabled << 3) | (static_cast<int>(device_refup_mode) << 2) | static_cast<int>(adcopt);
+    buffer_format[1] = (under_voltage_threshold & 0x0FF);
+    buffer_format[2] = ((over_voltage_threshold & 0x00F) << 4) | ((under_voltage_threshold & 0xF00) >> 8);
+    buffer_format[3] = ((over_voltage_threshold & 0xFF0) >> 4);
+    _start_wakeup_protocol();
+#ifdef USING_LTC6811_1
+    _write_config_through_broadcast(dcto_mode, buffer_format, cell_balance_statuses);
+#else
+    _write_config_through_address(dcto_mode, buffer_format, cell_balance_statuses);
+#endif
+}
+
+template <size_t num_chips, size_t num_chip_selects>
+void BMSDriverGroup<num_chips, num_chip_selects>::_write_config_through_broadcast(uint8_t dcto_mode, std::array<uint8_t, 6> buffer_format, const std::array<uint16_t, num_chips> &cell_balance_statuses)
+{
+    std::array<uint8_t, 4> cmd_and_pec;
+    std::array<uint8_t, (1 + num_chips) / 2 * 8> full_buffer;
```

Comment:
please be explicit with the order of operations using parenthesis for these template arguments. also it would be best if these were within a `constexpr` size variable for clarification.